### PR TITLE
feat(core): extract connectors as plugins

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

1. Extract connectors as plugins, locating at logto/packages/connector-xxx.
2. Wrap each connector as a class for future extendability.
3. Fix UTs according to refactor.
4. Fix imports and dependencies.

## Linear Issue Reference

LOG-2029
LOG-2236
LOG-2260

## Testing

Passed UTs.
Also tested connector functionalities:
image

### CODE REVIEW

File: packages/connector-aliyun-dm/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+      tsconfig: 'tsconfig.test.json',
+    },
+  },
+  moduleNameMapper: {
+    '^@/(.*)$': '<rootDir>/src/$1',
+    '^jose/(.*)$': '<rootDir>/node_modules/jose/dist/node/cjs/$1',
+  },
+  coveragePathIgnorePatterns: ['/node_modules/', '/build/'],
+  coverageReporters: ['text-summary', 'lcov'],
+};
```
Comment (IceHe) [inline packages/connector-aliyun-dm/jest.config.js:17]: Better write in `.ts`.
You can do it later.

Reference: `packages/ui/jest.config.ts`

File: packages/connector-facebook/src/index.test.ts [hunk 2]
```diff
@@ -11,33 +15,47 @@ const code = 'code';
...
 const fields = 'id,name,email,picture';
 
-jest.mock('../utilities');
+const mockedConfig = { clientId, clientSecret };
+const mockedTimeout = 5000;
+
+const getConnectorConfig = jest.fn() as GetConnectorConfig<FacebookConfig>;
+const getConnectorRequestTimeout = jest.fn() as GetTimeout;
+
+const FacebookMethods = new FacebookConnector(getConnectorConfig, getConnectorRequestTimeout);
```
Comment (IceHe) [inline packages/connector-facebook/src/index.test.ts:24]: nit: camelcase?

File: packages/connector-aliyun-sms/src/index.test.ts [hunk 3]
```diff
@@ -1,12 +1,19 @@
-import { sendMessage, validateConfig } from '.';
+import { GetConnectorConfig } from '@logto/connector-types';
+
+import { AliyunSmsConnector, AliyunSmsConfig } from '.';
 import { sendSms } from './single-send-text';
 
+const getConnectorConfig = jest.fn() as GetConnectorConfig<AliyunSmsConfig>;
+
+const AliyunSmsMethods = new AliyunSmsConnector(getConnectorConfig);
```
Comment (IceHe) [inline packages/connector-aliyun-sms/src/index.test.ts:8]: nit: camelcase

File: packages/connector-github/src/index.test.ts [hunk 4]
```diff
@@ -1,22 +1,37 @@
...
+const mockedConfig = {
+  clientId: '<client-id>',
+  clientSecret: '<client-secret>',
+};
+const mockedTimeout = 5000;
+
+const getConnectorConfig = jest.fn() as GetConnectorConfig<GithubConfig>;
+const getConnectorRequestTimeout = jest.fn() as GetTimeout;
+
+const GithubMethods = new GithubConnector(getConnectorConfig, getConnectorRequestTimeout);
```
Comment (IceHe) [inline packages/connector-github/src/index.test.ts:21]: nit: camelcase

File: packages/connector-google/src/index.test.ts [hunk 5]
```diff
@@ -1,37 +1,58 @@
...
+const mockedConfig = {
+  clientId: '<client-id>',
+  clientSecret: '<client-secret>',
+};
+const mockedTimeout = 5000;
+
+const getConnectorConfig = jest.fn() as GetConnectorConfig<GoogleConfig>;
+const getConnectorRequestTimeout = jest.fn() as GetTimeout;
+
+const GoogleMethods = new GoogleConnector(getConnectorConfig, getConnectorRequestTimeout);
```
Comment (IceHe) [inline packages/connector-google/src/index.test.ts:21]: nit: camelcase

File: packages/connector-wechat-native/src/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+const mockedConfig = {
+  appId: '<app-id>',
+  appSecret: '<app-secret>',
+};
+
+const getConnectorConfig = jest.fn() as GetConnectorConfig<WeChatConfig>;
+const getConnectorRequestTimeout = jest.fn() as GetTimeout;
+
+const WeChatNativeMethods = new WeChatNativeConnector(
```
Comment (IceHe) [inline packages/connector-wechat-native/src/index.test.ts:14]: nit: camelcase

File: packages/connector-wechat/src/index.test.ts [hunk 7]
```diff
@@ -1,22 +1,37 @@
...
+const mockedConfig = {
+  appId: '<app-id>',
+  appSecret: '<app-secret>',
+};
+const mockedTimeout = 5000;
+
+const getConnectorConfig = jest.fn() as GetConnectorConfig<WeChatConfig>;
+const getConnectorRequestTimeout = jest.fn() as GetTimeout;
+
+const WeChatMethods = new WeChatConnector(getConnectorConfig, getConnectorRequestTimeout);
```
Comment (IceHe) [inline packages/connector-wechat/src/index.test.ts:21]: nit: camelcase

File: packages/connector-alipay/src/index.test.ts [hunk 8]
```diff
@@ -14,32 +19,47 @@ import {
...
+  mockedTimestamp,
 } from './mock';
+import { AlipayConfig } from './types';
+
+const getConnectorConfig = jest.fn() as GetConnectorConfig<AlipayConfig>;
+const getConnectorRequestTimeout = jest.fn() as GetTimeout;
+const getConnectorTimestamp = jest.fn() as GetTimestamp;
 
-jest.mock('../utilities');
+const AlipayMethods = new AlipayConnector(
```
Comment (IceHe) [inline packages/connector-alipay/src/index.test.ts:31]: nit: camelcase

File: packages/connector-aliyun-dm/src/index.test.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import { GetConnectorConfig } from '@logto/connector-types';
+
+import { AliyunDmConnector, AliyunDmConfig } from '.';
+import { singleSendMail } from './single-send-mail';
+
+const getConnectorConfig = jest.fn() as GetConnectorConfig<AliyunDmConfig>;
+
+const AliyunDmMethods = new AliyunDmConnector(getConnectorConfig);
```
Comment (IceHe) [inline packages/connector-aliyun-dm/src/index.test.ts:8]: nit: camelcase

File: packages/connector-aliyun-sms/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  constructor(getConnectorConfig: GetConnectorConfig<AliyunSmsConfig>) {
+    this.getConfig = getConnectorConfig;
+  }
+
+  public validateConfig: ValidateConfig = async (config: unknown) => {
+    const result = configGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error.message);
+    }
```
Comment (IceHe) [inline packages/connector-aliyun-sms/src/index.ts:104]: assert

File: packages/connector-facebook/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    this.getConfig = getConnectorConfig;
+    this.getRequestTimeout = getConnectorRequestTimeout;
+  }
+
+  public validateConfig: ValidateConfig = async (config: unknown) => {
+    const result = facebookConfigGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error.message);
+    }
```
Comment (IceHe) [inline packages/connector-facebook/src/index.ts:80]: assert

File: packages/connector-github/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    this.getConfig = getConnectorConfig;
+    this.getRequestTimeout = getConnectorRequestTimeout;
+  }
+
+  public validateConfig: ValidateConfig = async (config: unknown) => {
+    const result = githubConfigGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error.message);
+    }
```
Comment (IceHe) [inline packages/connector-github/src/index.ts:75]: assert

File: packages/connector-google/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    this.getConfig = getConnectorConfig;
+    this.getRequestTimeout = getConnectorRequestTimeout;
+  }
+
+  public validateConfig: ValidateConfig = async (config: unknown) => {
+    const result = googleConfigGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error.message);
+    }
```
Comment (IceHe) [inline packages/connector-google/src/index.ts:80]: assert

File: packages/connector-wechat/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    this.getConfig = getConnectorConfig;
+    this.getRequestTimeout = getConnectorRequestTimeout;
+  }
+
+  public validateConfig: ValidateConfig = async (config: unknown) => {
+    const result = weChatConfigGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error.message);
+    }
```
Comment (IceHe) [inline packages/connector-wechat/src/index.ts:80]: assert

File: packages/connector-alipay/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+    this.getRequestTimeout = getConnectorRequestTimeout;
+    this.getTimestamp = getConnectorTimestamp;
+  }
+
+  public validateConfig: ValidateConfig = async (config: unknown) => {
+    const result = alipayConfigGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error.message);
+    }
```
Comment (IceHe) [inline packages/connector-alipay/src/index.ts:92]: assert

File: packages/connector-aliyun-dm/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  constructor(getConnectorConfig: GetConnectorConfig<AliyunDmConfig>) {
+    this.getConfig = getConnectorConfig;
+  }
+
+  public validateConfig: ValidateConfig = async (config: unknown) => {
+    const result = configGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error.message);
+    }
```
Comment (IceHe) [inline packages/connector-aliyun-dm/src/index.ts:81]: assert

File: packages/connector-types/package.json [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+  "author": "Logto Team",
+  "license": "MPL-2.0",
+  "files": [
+    "lib"
+  ],
+  "private": true,
+  "scripts": {
+    "preinstall": "npx only-allow pnpm",
+    "precommit": "lint-staged",
+    "build": "rm -rf lib/ && tsc --p tsconfig.build.json",
```
Comment (IceHe) [inline packages/connector-types/package.json:14]: `lib/` should be `build/`

See `packages/connector-*/tsconfig.base.json`

```json
{
  "extends": "@silverhand/ts-config/tsconfig.base",
  "compilerOptions": {
    "outDir": "build",
  ……
```

File: packages/connector-alipay/jest.config.js [hunk 18]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-alipay/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-aliyun-sms/jest.config.js [hunk 19]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-aliyun-sms/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-facebook/jest.config.js [hunk 20]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-facebook/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-github/jest.config.js [hunk 21]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-github/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-google/jest.config.js [hunk 22]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-google/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-types/jest.config.js [hunk 23]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-types/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-wechat-native/jest.config.js [hunk 24]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-wechat-native/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-wechat/jest.config.js [hunk 25]
```diff
@@ -0,0 +1,17 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (IceHe) [inline packages/connector-wechat/jest.config.js:1]: Better write in `.ts`.
You can do it later.

Reference: packages/ui/jest.config.ts

File: packages/connector-alipay/src/index.ts [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+      en: 'Sign In with Alipay',
+      'zh-CN': '支付宝登录',
+    },
+    // TODO: add the real logo URL (LOG-1823)
+    logo: './logo.png',
+    description: {
+      en: 'Sign In with Alipay',
+      'zh-CN': '支付宝登录',
+    },
+    readme: existsSync(pathToReadmeFile)
```
Comment (wangsijie) [inline packages/connector-alipay/src/index.ts:65]: can wrap this function in a package like `@logto/connector-utils`

File: packages/cli/package.json [hunk 27]
```diff
@@ -40,6 +40,7 @@
   },
   "prettier": "@silverhand/eslint-config/.prettierrc",
   "dependencies": {
+    "@logto/connector-types": "^0.1.0",
```
Comment (wangsijie) [inline packages/cli/package.json:43]: Does `cli` need this package?

File: packages/core/src/connectors/index.ts [hunk 28]
```diff
@@ -1,25 +1,70 @@
...
+import { getConnectorConfig, getConnectorRequestTimeout, getFormattedDate } from './utilities';
+
+const getAlipayConfig = async (id: string) => getConnectorConfig<AlipayConfig>(id);
+const AlipayIConnector = new AlipayConnector(
+  getAlipayConfig,
+  getConnectorRequestTimeout,
+  getFormattedDate
+) as SocialConnector;
+
+const getAliyunDmConfig = async (id: string) => getConnectorConfig<AliyunDmConfig>(id);
```
Comment (wangsijie) [inline packages/core/src/connectors/index.ts:24]: We don't need this function, just pass in "getConnectorConfig", the `aliyun-dm` it self can handle the typings by "T".

Comment (wangsijie) [reply packages/core/src/connectors/index.ts:24]: And then, `core` don't need to know the config type of specific connector.

**Review by wangsijie — COMMENTED** [review]
Overall looks OK to me, only need some fixes.

File: packages/connector-types/package.json (hunk 17 above)
Comment (gao-sun) [reply packages/connector-types/package.json]: this should be `lib/` actually?

**Pull request closed without merging**